core/embed/rust/src/trezorhal/display.rs

```diff
@@ -74,51 +80,51 @@ pub fn text_baseline(font: i32) -> i16 {
     unsafe { ffi::font_baseline(font).try_into().unwrap_or(i16::MAX) }
 }
 
-pub fn bar(x: i16, y: i16, w: i16, h: i16, fgcolor: u16) {
-    unsafe { ffi::display_bar(x.into(), y.into(), w.into(), h.into(), fgcolor) }
-}
-
-pub fn bar_radius(x: i16, y: i16, w: i16, h: i16, fgcolor: u16, bgcolor: u16, radius: u8) {
+#[inline(always)]
+#[cfg(all(feature = "disp_i8080_16bit_dw", not(feature = "disp_i8080_8bit_dw")))]
+pub fn pixeldata(c: u16) {
     unsafe {
-        ffi::display_bar_radius(
-            x.into(),
-            y.into(),
-            w.into(),
-            h.into(),
-            fgcolor,
-            bgcolor,
-            radius,
-        )
+        ffi::DISPLAY_DATA_ADDRESS.write_volatile(c);
     }
 }
 
-pub fn bar_radius_buffer(x: i16, y: i16, w: i16, h: i16, radius: u8, buffer: &mut BufferText) {
+#[cfg(feature = "framebuffer")]
+pub fn get_fb_addr() -> u32 {
+    unsafe { ffi::display_get_fb_addr() as u32 }
+}
+
+#[inline(always)]
+#[cfg(feature = "disp_i8080_8bit_dw")]
+pub fn pixeldata(c: u16) {
     unsafe {
-        ffi::display_bar_radius_buffer(
-            x.into(),
-            y.into(),
-            w.into(),
-            h.into(),
-            radius,
-            buffer.deref_mut(),
-        )
+        ffi::DISPLAY_DATA_ADDRESS.write_volatile((c & 0xff) as u8);
+        ffi::DISPLAY_DATA_ADDRESS.write_volatile((c >> 8) as u8);
     }
 }
 
 #[inline(always)]
-#[cfg(all(feature = "disp_i8080_16bit_dw", not(feature = "disp_i8080_8bit_dw")))]
-pub fn pixeldata(c: u16) {
+#[cfg(all(feature = "framebuffer", not(feature = "framebuffer32bit")))]
+pub fn pixel(fb: u32, x: i16, y: i16, c: u16) {
     unsafe {
-        ffi::DISPLAY_DATA_ADDRESS.write_volatile(c);
+        let addr = fb
+            + ((y as u32 + DISPLAY_FRAMEBUFFER_OFFSET_Y) * DISPLAY_FRAMEBUFFER_WIDTH
+                + (x as u32 + DISPLAY_FRAMEBUFFER_OFFSET_X))
+                * 2;
+        let ptr = addr as *mut u16;
+        ptr.write_volatile(c);
```

Comment:
I think the unsafe block only needs to be around this line?

Reply:
using offset function now on the pointer, so that needs unsafe too: https://github.com/trezor/trezor-firmware/commit/c37f89d7113ecd773327d2840223e44e7edd829c